### PR TITLE
Feature:sudo support for scp

### PR DESCRIPTION
**Pull request by ghostloda**
Add sudo support for scp  https://github.com/labring/sealos/issues/3402.

After this, wen can use sudo user to install k8s
```
sudo sealos run labring/kubernetes:v1.26.1 labring/helm:v3.10.3 labring/calico:v3.25.0 --masters 10.29.13.21 --nodes 10.29.13.22 -u sudouser -p xxx
```

TEST scp command
image

image

### CODE REVIEW

File: pkg/ssh/suftp.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+const promptpwd string = "gimmeyourpwdnow"
```
Comment (fanux) [inline pkg/ssh/suftp.go:29]: “give me your password now” ，Add spaces for better readability.

File: pkg/ssh/suftp.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+func NewSudoSftpClient(conn *ssh.Client, sudopwd string, opts ...sftp.ClientOption) (*sftp.Client, error) {
+	s, err := conn.NewSession()
+	if err != nil {
+		return nil, err
+	}
+
+	// serverpaths are the most likely paths to the sftp-server binary, ordered from most likely less likely
+	// paths are from https://winscp.net/eng/docs/faq_su#fn2
+	serverpaths := []string{
```
Comment (fanux) [inline pkg/ssh/suftp.go:42]: If the FTP server is not at this path, will there be compatibility issues?

**Review by fanux — APPROVED** [review]

File: pkg/ssh/suftp.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+const promptpwd string = "gimmeyourpwdnow"
+const promptsuccess string = "thesudopwdwasok"
+const prompterror string = "ohmgsudofailed!"
+const promptlength int = 15
+
+func NewSudoSftpClient(conn *ssh.Client, sudopwd string, opts ...sftp.ClientOption) (*sftp.Client, error) {
```
Comment (cuisongliu) [inline pkg/ssh/suftp.go:34]: check serverpaths is exsits?

**Review by cuisongliu — CHANGES_REQUESTED** [review]

File: pkg/ssh/suftp.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+	// sudo -p specifies the password prompt from sudo. The prompt is sent to stderr
+	// sudo -S dictates that password should be read from stdin
+	// sudo -u specifies the user
+	// sh -c 'cmd' passes cmd to sh
+	// >&2 echo "%s" echos promptsucess to stderr before starting sftp server ( see https://stackoverflow.com/a/23550347 )
+	// & %s is the final command to be executed is sudo is successful, which is the path to the sftp-server binary. It is a list of possible binary paths and it will pick the first one that exists. At the end, it will look in $PATH.
+	// || sh -c '>&2 echo "%s" & exit 1' -> runs if sudo failed and ensures that prompterror is written to stderr before erroring
+	// In total this means that we can read from stderr and write to stdin to
+	cmd := fmt.Sprintf(
```
Comment (fengxsong) [inline pkg/ssh/suftp.go:59]: Using this approach may be enough https://github.com/sealerio/sealer/blob/main/utils/ssh/connect.go#L195

btw, do we really need to support ask-pass scenario?

File: pkg/ssh/suftp.go (hunk 3 above)
Comment (ghostloda) [reply pkg/ssh/suftp.go:37]: replaced with **grep serverpaths  from '/etc/ssh/sshd_config'** , if file not found return error.

File: pkg/ssh/suftp.go (hunk 4 above)
Comment (ghostloda) [reply pkg/ssh/suftp.go:56]: Most of the production environments do not grant sudo user NOPASSWD privileges.

File: pkg/ssh/suftp.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+func getSftpServerPath(conn *ssh.Client, sudopwd string) (string, error) {
+	ses, err := conn.NewSession()
+	if err != nil {
+		return "", err
+	}
+	defer ses.Close()
+
+	cmd := fmt.Sprintf(`sudo -S grep -oP "Subsystem\s+sftp\s+\K.*" %s`, sshdConfigPath)
```
Comment (fengxsong) [inline pkg/ssh/suftp.go:151]: sudo is not required for this command

File: pkg/ssh/suftp.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+	sshdConfigPath        = "/etc/ssh/sshd_config"
+)
+
+func NewSudoSftpClient(conn *ssh.Client, sudopwd string, opts ...sftp.ClientOption) (*sftp.Client, error) {
+	s, err := conn.NewSession()
+	if err != nil {
+		return nil, err
+	}
+
+	sftpServerPath, err := getSftpServerPath(conn, sudopwd)
```
Comment (fengxsong) [inline pkg/ssh/suftp.go:43]: and perhaps here should use a cache, we don't want to get sftp-server path from remote ssh client each time when creating sftp session.

File: pkg/ssh/suftp.go (hunk 5 above)
Comment (ghostloda) [reply pkg/ssh/suftp.go:151]: > sudo is not required for this command

```
[root@node1 ~]# ll /etc/ssh/sshd_config 
-rw-------. 1 root root 3905 Aug  3  2022 /etc/ssh/sshd_config
```

File: pkg/ssh/suftp.go (hunk 6 above)
Comment (ghostloda) [reply pkg/ssh/suftp.go:43]: > and perhaps here should use a cache, we don't want to get sftp-server path from remote ssh client each time when creating sftp session.

Is sftp-client caching better?

File: pkg/ssh/suftp.go (hunk 5 above)
Comment (fengxsong) [reply pkg/ssh/suftp.go:151]: weird. the permission of this file on most of our servers are 644.

File: pkg/ssh/suftp.go (hunk 6 above)
Comment (fengxsong) [reply pkg/ssh/suftp.go:43]: of course yes, but most of the methods needs refactored.

Comment (ghostloda) [reply pkg/ssh/suftp.go:43]: Is it enough to add a cache under this function?

```
func (c *Client) newClientAndSftpClient(host string) (*ssh.Client, *sftp.Client, error) {
	var (
		sshClient  *ssh.Client
		sftpClient *sftp.Client
		err        error
	)
	sshClient, err = c.connect(host)
	if err != nil {
		return nil, nil, err
	}
	// create sftp client
	if c.Option.sudo || c.Option.user != defaultUsername {
		sftpClient, err = NewSudoSftpClient(sshClient, c.password)
	} else {
		sftpClient, err = sftp.NewClient(sshClient)
	}

	return sshClient, sftpClient, err
}
```

Comment (ghostloda) [reply pkg/ssh/suftp.go:43]: How about add a new pr to add a client cache? Like this approach https://github.com/sealerio/sealer/blob/main/utils/ssh/connect.go#L142

**Review by fanux — APPROVED** [review]

**Review by cuisongliu — APPROVED** [review]

**Pull request merged**